includes/admin/home/class-sensei-home-remote-data-provider.php

```diff
@@ -0,0 +1,133 @@
+<?php
+/**
+ * File containing Sensei_Home_Remote_Data_Provider class.
+ *
+ * @package sensei-lms
+ * @since   $$next-version$$
+ */
+
+/**
+ * Class responsible for fetching data from SenseiLMS.com.
+ */
+class Sensei_Home_Remote_Data_Provider {
+	const API_BASE_URL     = 'https://senseilms.com/wp-json/senseilms-home/1.0/';
+	const CACHE_KEY_PREFIX = 'sensei_home_remote_data_';
+	const CACHE_TTL        = DAY_IN_SECONDS;
+
+	/**
+	 * The primary plugin to pull data for.
+	 *
+	 * @var string
+	 */
+	private $primary_plugin_slug;
+
+	/**
+	 * Sensei_Home_Remote_Data_Provider constructor.
+	 *
+	 * @param string $primary_plugin_slug The primary plugin slug.
+	 */
+	public function __construct( string $primary_plugin_slug ) {
+		$this->primary_plugin_slug = $primary_plugin_slug;
+	}
+
+	/**
+	 * Fetch data from SenseiLMS.com.
+	 *
+	 * @param int $max_age Maximum age of the cached data in seconds. Max is 1 day (in seconds).
+	 *
+	 * @return array|\WP_Error
+	 */
+	public function fetch( int $max_age = null ) {
+		$url           = $this->get_api_url();
+		$transient_key = self::CACHE_KEY_PREFIX . md5( $url );
+		$data          = get_transient( $transient_key );
+
+		// If the data is too old, fetch it again.
+		if ( $max_age && is_array( $data ) ) {
+			$age = time() - ( $data['_fetched'] ?? 0 );
+			if ( $age > $max_age ) {
+				$data = false;
+			}
+
+			unset( $data['_fetched'] );
+		}
+
+		if ( false === $data ) {
+			$response = wp_safe_remote_get( $url );
+
+			if ( is_wp_error( $response ) ) {
+				return $response;
+			}
+
+			$data = json_decode( wp_remote_retrieve_body( $response ), true );
+			if ( ! is_array( $data ) ) {
+				return new WP_Error( 'sensei-home-remote-data-invalid-response' );
```

Comment:
I am a little bit concerned about flooding the SenseiLMS.com API in case an error is being generated. Should we also "persist" the error (maybe for a shorter period of time) so we don't do a huge amount of requests here?

Reply:
Good thought. I'll have a think about how we can do this. We might want to have a different retry strategy based on if it is for badges (any WP admin page) or for Sensei Home.

Reply:
I tried to implement this in https://github.com/Automattic/sensei/pull/5820/commits/79b657bee3068f73b02611cfd07746e702a63450